src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityAnalyzer.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Common;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class ImmutabilityAnalyzer : DiagnosticAnalyzer {
+
+		public const string DiagnosticId = "D2L0003";
+		private const string Category = "Safety";
+
+		private const string Title = "Classes marked as immutable should be immutable.";
+		private const string Description = "Classes marked as immutable or that implement interfaces marked immutable should be immutable.";
+		internal const string MessageFormat = "This class is marked immutable, but it is not. Check that all fields and properties are immutable.";
+
+		private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+			DiagnosticId, 
+			Title, 
+			MessageFormat, 
+			Category,
+			DiagnosticSeverity.Error,
+			isEnabledByDefault: true,
+			description: Description
+		);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create( Rule );
+
+		private readonly MutabilityInspector m_immutabilityInspector = new MutabilityInspector();
+
+		public override void Initialize( AnalysisContext context ) {
+			context.RegisterCompilationStartAction( RegisterIfNotTestAssembly );
+		}
+
+		private void RegisterIfNotTestAssembly( CompilationStartAnalysisContext compilation ) {
+			var references = compilation.Compilation.ReferencedAssemblyNames;
+			if( references.Any( r => r.Name.ToUpper().Contains( "NUNIT" ) ) ) {

```

Comment:
I think we're going to want to move away from this sorta thing in awhile but let's talk about that when we're both in the office